cirq-google/cirq_google/workflow/quantum_runtime.py

```diff
@@ -138,22 +178,63 @@ def __repr__(self) -> str:
         return _compat.dataclass_repr(self, namespace='cirq_google')
 
 
+def _safe_to_json(obj: Any, *, part_fn: str, nominal_fn: str, bak_fn: str):
+    """Safely update a json file.
+
+    1. The new value is written to a "part" file
+    2. The previous file atomically replaces the previous backup file, thereby becoming the
+       current backup file.
+    3. The part file is atomically renamed to the desired filename.
+    """
+    cirq.to_json_gzip(obj, part_fn)
+    if os.path.exists(nominal_fn):
+        os.replace(nominal_fn, bak_fn)
+    os.replace(part_fn, nominal_fn)
```

Comment:
Wouldn't it make sense to append the ".bak", ".part" extensions etc in here instead of requiring the caller to do that ?

Reply:
since these are all private functions, I don't think these sorts of issues are very important.

To have automated filename construction, you'd need to pass in the path's extensionless basename (or parse that from a normal filename). This could be error prone and might be overly restrictive if future (internal) code wants to use this function with filenames that don't necessarily follow the pattern.

I think being explicit here helps more than it hurts. And the stakes are low since these are all private APIs.